util/coverage/coverage_off_target_test.py

```diff
@@ -0,0 +1,237 @@
+        libs = list(set(libs_all) - set(libs_incompat))
+        # (1) return all libs, (2) return incompat., (3) build
+        mock_run.side_effect = (libs_all, libs_incompat, [])
+
+        self.assertEqual(coverage_off_target.gather_and_build_libs(), libs)
+
+        calls = mock_run.call_args_list
+        self.assertEqual(len(calls), 3)
+        self.assertEqual(calls[0][0][:2], (coverage_off_target.BAZEL, "query"))
+        self.assertEqual(calls[1][0][:2], (coverage_off_target.BAZEL, "cquery"))
+        self.assertEqual(calls[2][0], (coverage_off_target.BAZEL, "build",
+                                       "--config=coverage_clang", *libs))
+
+    @patch("coverage_off_target.run")
+    def test_create_merged_library(self, mock_run):
+        libs = ["//foo/bar1", "//foo/bar2"]
+        libs_query = "(//foo/bar1 + //foo/bar2)"
+        merged_so = Path("merged.so")
+        objs = ["foo/bar1.o", "foo/bar2.o"]
+        # (1) return object files, (2) link.
+        mock_run.side_effect = (objs, [])
+
+        coverage_off_target.create_merged_library(libs, merged_so)
+
+        calls = mock_run.call_args_list
+        self.assertEqual(len(calls), 2)
+        self.assertEqual(calls[0][0][:3],
+                         (coverage_off_target.BAZEL, "cquery", libs_query))
+        self.assertEqual(
+            calls[1][0],
+            (coverage_off_target.LLD, "--shared", "-o", str(merged_so), *objs))
+
+    @patch("coverage_off_target.run")
+    def test_measure_unit_test_coverage(self, mock_run):
+        tests = ["//foo:test1", "//foo:test2"]
+        testlogs = ["/some/long/path"]
+        profiles = [
+            "/some/long/path/foo/test1/coverage.dat",
+            "/some/long/path/foo/test2/coverage.dat"
+        ]
+        merged_profile = Path("merged.profdata")
+        # (1) find test targets, (2) measure coverage, (3) get testlogs dir, (4) merge
+        # profile data.
+        mock_run.side_effect = (tests, [], testlogs, [])
+
+        coverage_off_target.measure_unit_test_coverage(merged_profile)
+
+        calls = mock_run.call_args_list
+        self.assertEqual(len(calls), 4)
+        self.assertEqual(calls[0][0][:2], (coverage_off_target.BAZEL, "query"))
+        self.assertEqual(calls[1][0], (coverage_off_target.BAZEL, "coverage",
+                                       "--config=coverage_clang", *tests))
+        self.assertEqual(calls[2][0],
+                         (coverage_off_target.BAZEL, "info", "bazel-testlogs"))
+        self.assertEqual(
+            calls[3][0],
+            (coverage_off_target.LLVM_PROFDATA, "merge", "--sparse", "--output",
+             str(merged_profile), *profiles))
+
+    @patch("coverage_off_target.print")
+    @patch("pathlib.Path.open", new_callable=mock_open)
+    @patch("pathlib.Path.__truediv__",
+           side_effect=Path.__truediv__,
+           autospec=True)
+    @patch("coverage_off_target.run")
+    def test_generate_report(self, mock_run, mock_path_div, mock_path_open,
```

Comment:
~~Are these params out of order or do they need to be the reverse of the `@patch` lines?~~ Nevermind, looks like this is correct from looking at examples. https://docs.python.org/3/library/unittest.mock.html

Reply:
yes, they should be from the bottom up ([link to highlight](https://docs.python.org/3/library/unittest.mock.html#:~:text=Note%20When%20you%20nest%20patch%20decorators%20the%20mocks%20are%20passed%20in%20to%20the%20decorated%20function%20in%20the%20same%20order%20they%20applied%20(the%20normal%20Python%20order%20that%20decorators%20are%20applied).%20This%20means%20from%20the%20bottom%20up%2C%20so%20in%20the%20example%20above%20the%20mock%20for%20module.ClassName1%20is%20passed%20in%20first.)).